mlflow/bedrock/_autolog.py

```diff
@@ -0,0 +1,119 @@
+import io
+import json
+from typing import Any, Callable, Union
+
+from botocore.client import BaseClient
+from botocore.eventstream import EventStream
+
+import mlflow
+from mlflow.bedrock import FLAVOR_NAME
+from mlflow.entities import SpanType
+from mlflow.utils.autologging_utils import safe_patch
+from mlflow.utils.autologging_utils.config import AutoLoggingConfig
+
+_BEDROCK_RUNTIME_SERVICE_NAME = "bedrock-runtime"
+_BEDROCK_SPAN_PREFIX = "BedrockRuntime."
+
+
+def _skip_if_trace_disabled(func: Callable[..., Any]) -> Callable[..., Any]:
+    """
+    A decorator to apply the function only if trace autologging is enabled.
+    This decorator is used to skip the test if the trace autologging is disabled.
+    """
+
+    def wrapper(original, self, *args, **kwargs):
+        config = AutoLoggingConfig.init(flavor_name=FLAVOR_NAME)
+        if not config.log_traces:
+            return original(self, *args, **kwargs)
+
+        return func(original, self, *args, **kwargs)
+
+    return wrapper
+
+
+def patched_create_client(original, self, *args, **kwargs):
+    """
+    Patched version of the boto3 ClientCreator.create_client method that returns
+    a patched client class.
+    """
+    if kwargs.get("service_name") != _BEDROCK_RUNTIME_SERVICE_NAME:
+        return original(self, *args, **kwargs)
+
+    client = original(self, *args, **kwargs)
+    patch_bedrock_runtime_client(client.__class__)
+
+    return client
+
+
+def patch_bedrock_runtime_client(client_class: type[BaseClient]):
+    """
+    Patch the BedrockRuntime client to log traces and models.
+    """
+    # The most basic model invocation API
+    safe_patch(FLAVOR_NAME, client_class, "invoke_model", _patched_invoke_model)
+
+    if hasattr(client_class, "converse"):
+        # The new "converse" API was introduced in boto3 1.35 to access all models
+        # with the consistent chat format.
+        # https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/bedrock-runtime/client/converse.html
+        safe_patch(FLAVOR_NAME, client_class, "converse", _patched_converse)
+
+
+@_skip_if_trace_disabled
+def _patched_invoke_model(original, self, *args, **kwargs):
+    with mlflow.start_span(name=f"{_BEDROCK_SPAN_PREFIX}{original.__name__}") as span:
+        # NB: Bedrock client doesn't accept any positional arguments
+        span.set_inputs(kwargs)
+
+        result = original(self, *args, **kwargs)
+
+        result["body"] = _buffer_stream(result["body"])
+        parsed_response_body = _parse_invoke_model_response_body(result["body"])
```

Comment:
in the screenshot on the PR description, it looks like `body` is still a string, but reading the code it looks like it should be dict after this line. is there something going wrong?

Reply:
Oh the `body` shown in the screenshot is the one in request, not the response. The one in the response is extracted as a dictionary:

![Screenshot 2024-12-18 at 14 35 04](https://github.com/user-attachments/assets/94a6e766-52ee-4ca6-b15d-2b47615be60c)